### ISSUE TITLE
相似句生成的问题

### ISSUE THREAD

**Issue by Yi-Shi94** [JunnYu/RoFormer_pytorch#26]
第一次接触seq2seq和相似句生成。
我按照 #17 里提到的使用RoFormerForMaskedLM，设config.is_decoder=True，参考UniLM_Mask，替换掉了对应的位置
```
def get_extended_attention_mask(self, seg_id):
        idxs = torch.cumsum(seg_id,dim=1)
        mask = idxs[:, None, :] <= idxs[:, :, None]
        mask =  1.0 * mask[:,None] #-(1.0 - mask[:, None]) * 1000000.0
        return mask
```
但输出的结果是不对的。

我现在的后续实现：将  [CLS]你叫什么？[SEP] 输入模型；把[SEP]位置的logit拿出求出概率，使用greedysearch拿出最大概率的id，加入input_ids；同时更新seg_id以求出新的 extended attention mask；对于最后一位logit重复之前操作直到输出终止符或超出长度。
请问哪里出问题了？@[hxs91](https://github.com/hxs91)@[JunnYu](https://github.com/JunnYu)

**Comment by JunnYu** [comment]
有完整测试的代码吗，有的话麻烦提供一下，谢谢
猜测可能的问题是，在generate过程中Q,K并未施加上旋转位置编码。

**Comment by JunnYu** [comment]
# 如果想要生成结果的话，下载这下面的roformer就可以生成结果了。
# [roformer.zip](https://github.com/JunnYu/RoFormer_pytorch/files/8402725/roformer.zip)

```python
import torch
import numpy as np
from roformer import RoFormerTokenizer, RoFormerForCausalLM, RoFormerConfig

device = torch.device('cuda:0' if torch.cuda.is_available() else 'cpu')
pretrained_model = "junnyu/roformer_chinese_sim_char_base"
tokenizer = RoFormerTokenizer.from_pretrained(pretrained_model)
config = RoFormerConfig.from_pretrained(pretrained_model)
config.is_decoder = True
config.eos_token_id = tokenizer.sep_token_id
config.pooler_activation = "linear"
model = RoFormerForCausalLM.from_pretrained(pretrained_model, config=config)
model.to(device)
model.eval()

def gen_synonyms(text, n=100, k=20):
    ''''含义： 产生sent的n个相似句，然后返回最相似的k个。
    做法：用seq2seq生成，并用encoder算相似度并排序。
    '''
    # 寻找所有相似的句子
    r = []
    inputs1 = tokenizer(text, return_tensors="pt")
    for _ in range(n):
        inputs1.to(device)
        output = tokenizer.batch_decode(model.generate(**inputs1, top_p=0.95, do_sample=True, max_length=128), skip_special_tokens=True)[0].replace(" ","").replace(text, "") # 去除空格，去除原始text文本。
        r.append(output)
    
    # 对相似的句子进行排序
    r = [i for i in set(r) if i != text and len(i) > 0]
    r = [text] + r
    inputs2 = tokenizer(r, padding=True, return_tensors="pt")
    with torch.no_grad():
        inputs2.to(device)
        outputs = model(**inputs2)
        Z = outputs.pooler_output.cpu().numpy()
    Z /= (Z**2).sum(axis=1, keepdims=True)**0.5
    argsort = np.dot(Z[1:], -Z[0]).argsort()
    
    return [r[i + 1] for i in argsort[:k]]

out = gen_synonyms("广州和深圳哪个好？")
print(out)
# ['深圳和广州哪个好？',
#  '广州和深圳哪个好',
#  '深圳和广州哪个好',
#  '深圳和广州哪个比较好。',
#  '深圳和广州哪个最好？',
#  '深圳和广州哪个比较好',
#  '广州和深圳那个比较好',
#  '深圳和广州哪个更好？',
#  '深圳与广州哪个好',
#  '深圳和广州，哪个比较好',
#  '广州与深圳比较哪个好',
#  '深圳和广州哪里比较好',
#  '深圳还是广州比较好？',
#  '广州和深圳哪个地方好一些？',
#  '广州好还是深圳好？',
#  '广州好还是深圳好呢？',
#  '广州与深圳哪个地方好点？',
#  '深圳好还是广州好',
#  '广州好还是深圳好',
#  '广州和深圳哪个城市好？']
```

**Comment by Yi-Shi94** [comment]
解决了，非常感谢。
